### PR TITLE
fix(evm): precompile evm gas use based on input, logic from openethereum

### PR DESCRIPTION
**Pull request by ailisp**
Address #3552. neard,  runtime_params_estimator and tests okay.
measure evm precompile cost is removed from runtime-param-estimator, using numbers in openethereum/ethcore/res/ethereum/ethercore.json because
- openethereum gives evm gas usage for both "precompile functions" and "all other operations"
- assume they're compatible and correct (This is a fair assumption because as an ethereum client, openethereum should give an accurate evm gas usage, in both with or without precompile functions)
- so there should be no specific "evm precompile function" gas estimation in runtime_params_estimator, still rely on "function call near base cost" + "function call near cost per evm"

Test Plan
-------------
standard_ops tests. runtime_params_estimator compiles and works

### CODE REVIEW

File: runtime/near-evm-runner/src/builtins.rs [hunk 1]
```diff
@@ -598,4 +643,426 @@ impl Impl for Blake2FImpl {
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Open Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+/// A gas pricing scheme for built-in contracts.
+trait Pricer: Send + Sync {
+    /// The gas cost of running this built-in for the given input data at block number `at`
+    fn cost(&self, input: &[u8]) -> U256;
+}
```
Comment (evgenykuzyakov) [inline runtime/near-evm-runner/src/builtins.rs:672]: Can you move `Pricer` trait and implementations to a separate mod?

Comment (ailisp) [reply runtime/near-evm-runner/src/builtins.rs]: done

File: runtime/near-evm-runner/src/pricer.rs [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Open Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use eth_pairings::public_interface::eip2537::{
+    SCALAR_BYTE_LENGTH, SERIALIZED_G1_POINT_BYTE_LENGTH, SERIALIZED_G2_POINT_BYTE_LENGTH,
+};
```
Comment (evgenykuzyakov) [inline runtime/near-evm-runner/src/pricer.rs:19]: Let's just copy the constants here instead of creating a dependency.
We can attribute to the source

Comment (ailisp) [reply runtime/near-evm-runner/src/pricer.rs:19]: Sounds good

File: runtime/near-evm-runner/src/pricer.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+            let len = min(exp_len, 32) as usize;
+            reader
+                .read_exact(&mut buf[(32 - len)..])
+                .expect("reading from zero-extended memory cannot fail; qed");
+            U256::from_big_endian(&buf[..])
+        };
+
+        let adjusted_exp_len = Self::adjusted_exp_len(exp_len, exp_low);
+
+        let (gas, overflow) = Self::mult_complexity(m).overflowing_mul(max(adjusted_exp_len, 1));
```
Comment (evgenykuzyakov) [inline runtime/near-evm-runner/src/pricer.rs:179]: FYI. No action needed.
This code is terrible and hard to read. A lot of overflow checks are just coming from the fact that some variable was limited in length. 
Why this is limited to `u64`? Is Ethereum maximum gas value limited to `u64`? I assume yes, so it's probably okay.

File: runtime/near-runtime-fees/src/lib.rs [hunk 4]
```diff
@@ -323,11 +363,17 @@ impl RuntimeFeesConfig {
...
-                sha256_cost: 0,
-                ripemd160_cost: 0,
-                identity_cost: 0,
-                modexp_cost: 0,
+                precompile_costs: EvmPrecompileCostConfig {
+                    ecrecover_cost: EvmLinearCost { base: 0, word: 0 },
+                    sha256_cost: EvmLinearCost { base: 0, word: 0 },
+                    ripemd160_cost: EvmLinearCost { base: 0, word: 0 },
+                    identity_cost: EvmLinearCost { base: 0, word: 0 },
+                    modexp_cost: EvmModexpCost { divisor: 0 },
```
Comment (evgenykuzyakov) [inline runtime/near-runtime-fees/src/lib.rs:371]: Can a divisor be `0`? Maybe it should be at least `1`, but then we need to handle free case.

I suggest we implement that if `divisor == 0` then gas `0` in `ModexpPricer`

Comment (ailisp) [reply runtime/near-runtime-fees/src/lib.rs:371]: Good point

File: runtime/near-evm-runner/src/pricer.rs (hunk 3 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/pricer.rs:179]: didn't find anywhere documented maximum ethereum gas, but i guess no one has enough ether to prepay more than u64::max amount of gas

**Pull request merged**